pkg/server/k8s/client.go

```diff
@@ -312,16 +311,133 @@ func (k *k8sClient) Limits(namespace, name string) (*app.Limits, error) {
+	replicas := k.currentPodReplicasFromDeploy(deploySpec.Namespace, deploySpec.Name)
+	deployYaml := deploySpecToK8sDeploy(deploySpec, replicas)
+	_, err := k.kc.Deployments(deploySpec.Namespace).Update(deployYaml)
+	if k.IsNotFound(err) {
+		_, err = k.kc.Deployments(deploySpec.Namespace).Create(deployYaml)
+	}
+	return err
+}
+
+func (k *k8sClient) PodRun(podSpec *deploy.PodSpec) (io.ReadCloser, <-chan int, error) {
+	podYaml := podSpecToK8sPod(podSpec)
+	pod, err := k.kc.Pods(podSpec.Namespace).Create(podYaml)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	exitCodeChan := make(chan int)
+	r, w := io.Pipe()
+	go func() {
+		defer func() {
+			w.Close()
+			close(exitCodeChan)
+		}()
+
+		if err := k.waitPodStart(pod, 1*time.Second, 5*time.Minute); err != nil {
+			return
+		}
+
+		stream, err := k.PodLogs(podSpec.Namespace, podSpec.Name, 10, true)
+		if err != nil {
+			return
+		}
+		io.Copy(w, stream)
+
+		if err = k.waitPodEnd(pod, 1*time.Second, 5*time.Minute); err != nil {
+			return
+		}
+
+		exitCodeChan <- k.podExitCode(pod)
+		go k.killPod(pod)
+	}()
+	return r, exitCodeChan, nil
+}
+
+func (k *k8sClient) HasService(namespace, appName string) (bool, error) {
+	_, err := k.kc.CoreV1().Services(namespace).Get(appName)
+	if err != nil {
+		if k.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (k *k8sClient) CreateService(namespace, appName string) error {
+	srvSpec := serviceSpec(namespace, appName, k.defaultServiceType)
+	_, err := k.kc.CoreV1().Services(namespace).Create(srvSpec)
+	return err
+}
+
+func (k *k8sClient) killPod(pod *k8sv1.Pod) error {
+	return k.kc.Pods(pod.Namespace).Delete(pod.Name, &k8sv1.DeleteOptions{})
+}
+
+func (k *k8sClient) waitPodStart(pod *k8sv1.Pod, checkInterval, timeout time.Duration) error {
+	podsClient := k.kc.Pods(pod.Namespace)
+	return wait.PollImmediate(checkInterval, timeout, func() (bool, error) {
+		p, err := podsClient.Get(pod.Name)
+		if err != nil {
+			return false, err
+		}
+		if p.Status.Phase == k8sv1.PodFailed {
+			return true, ErrPodRunFailed
+		}
+		result := p.Status.Phase == k8sv1.PodRunning || p.Status.Phase == k8sv1.PodSucceeded
+		return result, nil
+	})
+}
+
+func (k *k8sClient) waitPodEnd(pod *k8sv1.Pod, checkInterval, timeout time.Duration) error {
+	podsClient := k.kc.Pods(pod.Namespace)
+	return wait.PollImmediate(checkInterval, timeout, func() (bool, error) {
+		p, err := podsClient.Get(pod.Name)
+		if err != nil {
+			return false, err
+		}
+		result := p.Status.Phase == k8sv1.PodSucceeded || p.Status.Phase == k8sv1.PodFailed
+		return result, nil
+	})
+}
+
+func (k *k8sClient) podExitCode(pod *k8sv1.Pod) int {
+	p, err := k.kc.Pods(pod.Namespace).Get(pod.Name)
+	if err != nil {
+		return 1
+	}
+	for _, containerStatus := range p.Status.ContainerStatuses {
+		state := containerStatus.State.Terminated
+		if state.ExitCode != 0 {
+			return int(state.ExitCode)
+		}
+	}
+	return 0
```

Comment:
If I call this function on a running pod I would expect an error: pod is still running, no exit code yet. But this is :nail_care: